lightning/src/ln/interactivetxs.rs

```diff
@@ -349,19 +364,40 @@ impl NegotiationContext {
 			return Err(AbortReason::ExceededNumberOfInputsOrOutputs);
 		}
 
-		// TODO: How do we enforce their fees cover the witness without knowing its expected length?
-		const INPUT_WEIGHT: u64 = BASE_INPUT_WEIGHT + EMPTY_SCRIPT_SIG_WEIGHT;
-
 		// - the peer's paid feerate does not meet or exceed the agreed feerate (based on the minimum fee).
-		let counterparty_output_weight_contributed: u64 = self
+		let mut counterparty_weight_contributed: u64 = self
 			.counterparty_outputs_contributed()
 			.map(|output| {
 				(8 /* value */ + output.script_pubkey.consensus_encode(&mut sink()).unwrap() as u64)
 					* WITNESS_SCALE_FACTOR as u64
 			})
 			.sum();
-		let counterparty_weight_contributed = counterparty_output_weight_contributed
-			+ self.counterparty_inputs_contributed().count() as u64 * INPUT_WEIGHT;
+		// We don't know the counterparty's witnesses ahead of time obviously, so we need to have some
+		// estimates for their weights based on script type. We make some reasonably assumptions here.
+		let mut total_inputs_weight: u64 = 0;
+		for TxInputWithPrevOutput { prev_output, .. } in self.counterparty_inputs_contributed() {
+			total_inputs_weight = total_inputs_weight
+				.saturating_add(BASE_INPUT_WEIGHT)
+				.saturating_add(if prev_output.script_pubkey.is_v0_p2wpkh() {
+					// The weight for a P2WPKH spend is fixed.
+					EMPTY_SCRIPT_SIG_WEIGHT + P2WPKH_WITNESS_WEIGHT
+				} else if prev_output.script_pubkey.is_v0_p2wsh() {
+					// We cover a 2-of-2 multisig spend case as a minimum estimated weight. We assume signature
```

Comment:
Proposal: move out witness size estimation in a separate method. Also some unit tests for it would be nice.